sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/circuitBreaker/ConsecutiveExceptionBasedCircuitBreaker.java

```diff
@@ -0,0 +1,219 @@
+                        .withExceptionThresholdBreached(locationSpecificHealthContext.isExceptionThresholdBreached())
+                        .build();
+
+                } else {
+
+                    return new LocationSpecificHealthContext.Builder()
+                        .withSuccessCountForWriteForRecovery(successCountAfterHandling)
+                        .withExceptionCountForWriteForCircuitBreaking(exceptionCountAfterHandling)
+                        .withSuccessCountForReadForRecovery(locationSpecificHealthContext.getSuccessCountForReadForRecovery())
+                        .withExceptionCountForReadForCircuitBreaking(locationSpecificHealthContext.getExceptionCountForReadForCircuitBreaking())
+                        .withUnavailableSince(locationSpecificHealthContext.getUnavailableSince())
+                        .withLocationHealthStatus(locationSpecificHealthContext.getLocationHealthStatus())
+                        .withExceptionThresholdBreached(locationSpecificHealthContext.isExceptionThresholdBreached())
+                        .build();
+
+                }
+            case Unavailable:
+                throw new IllegalStateException();
+            default:
+                throw new IllegalArgumentException();
+        }
+    }
+
+    public boolean shouldHealthStatusBeDowngraded(LocationSpecificHealthContext locationSpecificHealthContext, boolean isReadOnlyRequest) {
+
+        int exceptionCountActual
+            = isReadOnlyRequest ? locationSpecificHealthContext.getExceptionCountForReadForCircuitBreaking() : locationSpecificHealthContext.getExceptionCountForWriteForCircuitBreaking();
+
+        return exceptionCountActual >= getAllowedExceptionCountToMaintainStatus(locationSpecificHealthContext.getLocationHealthStatus(), isReadOnlyRequest);
+    }
+
+    public boolean canHealthStatusBeUpgraded(LocationSpecificHealthContext locationSpecificHealthContext, boolean isReadOnlyRequest) {
+
+        int successCountActual
+            = isReadOnlyRequest ? locationSpecificHealthContext.getSuccessCountForReadForRecovery() : locationSpecificHealthContext.getSuccessCountForWriteForRecovery();
+
+        LocationHealthStatus locationHealthStatus = locationSpecificHealthContext.getLocationHealthStatus();
+
+        return successCountActual >= getMinimumSuccessCountForStatusUpgrade(locationHealthStatus, isReadOnlyRequest);
+    }
+
+    public int getAllowedExceptionCountToMaintainStatus(LocationHealthStatus status, boolean isReadOnlyRequest) {
+
+        if (isReadOnlyRequest) {
+            switch (status) {
+                case HealthyWithFailures:
+                    return this.partitionLevelCircuitBreakerConfig.getConsecutiveExceptionCountToleratedForReads();
+                case HealthyTentative:
+                    return this.partitionLevelCircuitBreakerConfig.getConsecutiveExceptionCountToleratedForReads() / 2;
```

Comment:
what if getConsecutiveExceptionCountToleratedForReads()  is 1? Does 2 need to be a config settting?

Reply:
ditto below for writes

Reply:
Worth discussing in our sync up on the guidance for this - idea was to make a partition health recovery conservative (twice as conservative here) but at the same time if exception count tolerated is 1 or 2 then we are failing over too quickly in the first place. This will increase latency due to cross-region calls and introduce behavior where we simply cycle through all regions as we keep failing over and back.

Reply:
Similar as my comment in Configs.java - I would start with hard lower/upper boundaries on the config settings - we can always relax them - but never make them stricter (breaking change).